Bugfix/s3 utils 188 support metrics above max safe

benzekrimaha opened the pull request.

bert-e commented:
# Hello benzekrimaha,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/s3utils/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-188 contains:


* `1.14.17`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.14.18`

* `1.15.8`


Please check the `Fix Version/s` of S3UTILS-188, or the target
branch of this pull request.

benzekrimaha commented:
/approve

benzekrimaha commented:
/wait

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-188 contains:


* `1.14.17`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.14.18`

* `1.15.8`


Please check the `Fix Version/s` of S3UTILS-188, or the target
branch of this pull request.




*The following options are set:* **approve**

Seragonia committed:
Update eslint config to support modern bigint notations

Issue: S3UTILS-188

Seragonia committed:
Update and correct dev dependencies

- Jest version was too old to support BigInts
- randomatic was used in some tests, yet never part of the project

Issue: S3UTILS-188

Seragonia committed:
Bump Arsenal

Issue: S3UTILS-188

Seragonia committed:
[maintenance] keep CI up to date

Issue: S3UTILS-188

Seragonia committed:
Update bucketVersionsStats to work with BigInts

Issue: S3UTILS-188

Seragonia committed:
Update VeeamSOSAPI consumption cronjob to support bigint

- We are backward compatible
- But we now store the values as Long in MongoDB, explicitly,
  as it would be the case with large values anyway.

Issue: S3UTILS-188

Seragonia committed:
Update consolidate metrics to accept number and bigints

- Output always bigint
- The conversion in place is to ensure backward compatibility:
  values might be numbers after an upgrade. It can be safely
  removed in an upcoming version once we confirm all customers
  are up to date.

Issue: S3UTILS-188

Seragonia committed:
Handle bigint communications internally

- BigInts are serialized with a __bigint field and passed as
  string.
- When the receiver gets the data, the value is automatically
  converted into.

Issue: S3UTILS-188

Seragonia committed:
Update Arsenal version

Issue: S3UTILS-188

Seragonia committed:
Relax eslint rule

- So we can follow the node naming convention for BigInt

Issue: S3UTILS-188

Seragonia committed:
Support storage consumption metrics as bigints

Issue: S3UTILS-188

Seragonia committed:
Update tests

Issue: S3UTILS-188

ghost commented:
The PR was updated to :
- Compute all metrics with bigints
- Store the storage metrics as Long in MongoDB
- Store the global metrics (in the document with `_id` `countitems`) as numbers, to not change anything. The values here will be imprecise but much more work is required if we want to support it properly everywhere.

williamlardier commented on package.json:
```diff
@@ -28,7 +28,7 @@
     "@senx/warp10": "^1.1.2",
     "JSONStream": "^1.3.5",
-    "arsenal": "git+https://github.com/scality/arsenal#8.1.142",
+    "arsenal": "git+https://github.com/scality/arsenal#d59d00d28b96748bd0099b50f6cfb84827a49dc6",
```
To be changed with a proper tag before merging

Seragonia committed:
Test with actual big numbers

Issue: S3UTILS-188

ghost commented:
PR updated with a unit test to ensure we compute the values properly when we have bigints. Too hard to test as functional test: we won't be uploading large objects in the CI.

williamlardier commented on CountItems/CountWorker.js:
```diff
@@ -17,6 +17,35 @@ class CountWorker {
         this.handleMessage = this.handleMessage.bind(this);
     }
 
+    _serializeBigInts(obj) {
```
The idea is that in s3utils we have a master-worker approach, where workers send their results to the master. This process serializes the results; yet the master has a need to know what values are bigint, and what are strings.

As such, these functions (and the equivalent ones in the master process class) will automatically use a special field, `__bigint`, to dynamically detect that an object is in fact a bigint.

williamlardier replied:
@benzekrimaha we might want to dedup this, with some static functions in a util file

Seragonia committed:
Fix the Veeam Available field being -1 if capacity is exceeded

Issue: S3UTILS-188

Seragonia committed:
Dedup (de)serializer functions

Issue: S3UTILS-188

Seragonia committed:
Bump project version

Issue: S3UTILS-188

francoisferrand commented on CountItems/utils/utils.js:
```diff
@@ -9,79 +9,111 @@ function consolidateDataMetrics(target, source) {
-    resTarget.usedCapacity._nonCurrentRestoring += usedCapacity && usedCapacity._nonCurrentRestoring ? usedCapacity._nonCurrentRestoring : 0;
-    resTarget.usedCapacity._nonCurrentRestored += usedCapacity && usedCapacity._nonCurrentRestored ? usedCapacity._nonCurrentRestored : 0;
-    resTarget.usedCapacity._incompleteMPUParts += usedCapacity && usedCapacity._incompleteMPUParts ? usedCapacity._incompleteMPUParts : 0;
+    resTarget.usedCapacity.current += BigInt(usedCapacity?.current || 0n);
```
not sure we should take the chance and remove these, but it seems to me these `BigInt` conversions should not be useful : since these fields are only managed by `_processEntryData()` (which already handles bigint) and this function...

williamlardier replied:
I tried, but we end up with functional tests failures, maybe @benzekrimaha you will have a bit more info on this?

francoisferrand replied:
is it worth investigating more, to catch issue or just leave the code cleaner?
(if you are confident it works well enough, but not happy with the code, we could merge "as is" and investigate after the release)

williamlardier replied:
I think it's safe and it's more like we don't pass the right values to this function in the test, so not in the actual program, having passing tests would be rethink them. Better be extra-safe here, as this is the most sensitive part of our changes.

francoisferrand commented on CountItems/utils/utils.js:
```diff
@@ -9,79 +9,111 @@ function consolidateDataMetrics(target, source) {
     return resTarget;
 }
 
+
```
nit: extra empty line

francoisferrand commented on CountItems/utils/utils.js:
```diff
@@ -9,79 +9,111 @@ function consolidateDataMetrics(target, source) {
 }
 
+
+function serializeBigInts(obj) {
```
nit: performance-wise, it is better to have such generic functions or custom functions for the specific objects we convert? (since we it is actually used for either `BucketInfo` or `BucketInfo[]`) ?

williamlardier replied:
As discussed I believe this is minor as ran once per bucket, and eventually it allows us to be flexible in the future, if we were to add more fields.

francoisferrand commented on tests/functional/utils/S3UtilsMongoClient.js:
```diff
@@ -1581,17 +1613,17 @@ describe('S3UtilsMongoClient::getObjectMDStats', () => {
 });
 
-describe('S3UtilsMongoClient::updateBucketCapacityInfo', () => {
+describe.skip('S3UtilsMongoClient::updateBucketCapacityInfo', () => {
```
test is skipped

francoisferrand commented on tests/functional/utils/S3UtilsMongoClient.js:
```diff
@@ -1644,17 +1676,17 @@ describe('S3UtilsMongoClient::updateBucketCapacityInfo', () => {
 });
 
-describe('S3UtilsMongoClient::getBucketInfos', () => {
+describe.skip('S3UtilsMongoClient::getBucketInfos', () => {
```
test is skipped

francoisferrand commented on tests/functional/collectBucketMetricsAndUpdateBucketCapacityInfo.js:
```diff
@@ -94,8 +94,8 @@ describe('collectBucketMetricsAndUpdateBucketCapacityInfo', () => {
             next => client.getBucketAttributes(testBucketName, logger, (err, bucketInfo) => {
-                assert.equal(err, null);
-                assert.equal(bucketInfo.getCapabilities(), null);
+                assert.equal(err && err.message, 'InternalError');
+                assert.equal(bucketInfo, undefined);
                 next();
```
seems weird to expect an `InternalError` on `getBucketAttributes()`...
(and it does not really to the test description)

williamlardier replied:
I updated the test, it was not the right check

williamlardier replied:
Interesting point though, it was passing for the wrong reasons: the assert in the callback, when failing, would never reach the `done`, but instead, throw n error that will be catched by whatever can do it; and here, it was in the mongoclientinterface clas... so we end up calling this test callback a second time, and with the right error.

Note that this is not the first time i see this and we have this problem in almost all our node codebase.

francoisferrand commented on utils/S3UtilsMongoClient.js:
```diff
@@ -104,20 +114,19 @@ class S3UtilsMongoClient extends MongoClientInterface {
                 if (id.startsWith('bucket_')) {
                     const inflightDocument = inflightsMap[id];
-                    const inflight = Long.fromNumber(Number(inflightDocument ? Math.max(0, inflightDocument - entry.usedCapacity._inflightsPreScan) : 0));
+                    const inflight = inflightDocument ? BigIntMax(0, inflightDocument - entry.usedCapacity._inflightsPreScan) : 0n;
```
```suggestion
                    const inflight = inflightDocument ? BigIntMax(0n, inflightDocument - entry.usedCapacity._inflightsPreScan) : 0n;
```

francoisferrand commented on utils/S3UtilsMongoClient.js:
```diff
@@ -13,37 +13,43 @@ const USERSBUCKET = '__usersbucket';
 const __COUNT_ITEMS = 'countitems';
 
+const BigIntMax = (...args) => args.reduce((max, current) => {
+    const maxAsBigInt = typeof max === 'bigint' ? max : BigInt(max);
+    const currentAsBigInt = typeof current === 'bigint' ? current : BigInt(current);
+    return maxAsBigInt > currentAsBigInt ? max : current;
```
* nit: same as earlier comment: it seems to be this type-check & BigInt conversion should not be needed...
* if in doubt, should probably just call BigInt constructor anyway, to be consistent (if there a perf impact we try to mitigate, then we should do it elsewhere as well...)

```suggestion
    const maxAsBigInt = BigInt(max);
    const currentAsBigInt = BigInt(current);
    return maxAsBigInt > currentAsBigInt ? max : current;
```

francoisferrand commented on utils/S3UtilsMongoClient.js:
```diff
@@ -754,11 +766,23 @@ class S3UtilsMongoClient extends MongoClientInterface {
         return {
-            versions: Math.max(0, totalNonCurrentCount + totalNonCurrentColdCount + totalVersionRestoringCount + totalVerionsRestoredCount),
-            objects: totalCurrentCount + totalCurrentColdCount + totalRestoringCount + totalRestoredCount,
+            // We do not support bigint for the global count items document
```
what limitation are we adressing?
* here we perform the computation, and it will be written to `Long` anyway in mongo : so should not affect countItems
* if we convert to Number eventually, might as well do all calculus in number...
* practically, there is no immediate risk: we are not close to the limit on the number of objects (our DB will probably break first)

e.g. I am fine to either do the computation in bigint or in number, but it's a waste to do it in BigInt to eventually convert to number while a Long is stored in Mongo...

williamlardier replied:
It's because we store these values as numbers today, and migrating this is not urgent here. Contrary to what I previously said, the driver does not enforce a type automatically, so even if we exceed the max safe size, we re still storing a number. The `Long` are coming from a previous change we made.

So the issue here is that we also store the sum of all the managed data by location, and this may be bigger than a number -> supporting long will require changes in pensieve-api, cloudserver, and possibly more? So it's not about avoiding the change, but reduce as much as possible the risks here.

Example:
```
  {
    _id: 'countitems',
    value: {
      objects: 105027,
      versions: 0,
      buckets: 1,
      bucketList: [
        ...
      ],
      dataManaged: {
        total: { curr: 115465213591495740, prev: 0 },
        byLocation: {
          '9a0c45b5-759a-11ef-9efd-52d84ee9ecd5': { curr: 115465213591495740, prev: 0 }
        }
      },
      stalled: 0
    }
  },
```

francoisferrand replied:
so we should add this as well in the list of "followup works"

Seragonia committed:
Improve code from review

Issue: S3UTILS-188

francoisferrand commented on tests/functional/utils/S3UtilsMongoClient.js:
```diff
@@ -451,7 +464,26 @@ describe('S3UtilsMongoClient::getObjectMDStats', () => {
                             (err, data) => {
                                 assert.deepStrictEqual(err, null);
-                                assert.deepStrictEqual(data, expected);
+                                // assert.deepStrictEqual(data, expected);\
```
commented-out code

williamlardier replied:
It was meant to be removed in favor of the below lines, so I removed it.

francoisferrand commented on tests/functional/collectBucketMetricsAndUpdateBucketCapacityInfo.js:
```diff
@@ -160,16 +160,16 @@ describe('collectBucketMetricsAndUpdateBucketCapacityInfo', () => {
-    test('should update bucket Capacity and Available -1 if Capacity value is not valid', done => {
-        testBucketCapacities.VeeamSOSApi.CapacityInfo.Capacity = 'not-a-number';
+    test('should update bucket Capacity and Available -1 if Capacity value is not set', done => {
+        testBucketCapacities.VeeamSOSApi.CapacityInfo.Capacity = -1;
```
should it be `-1n` ?
or is it the case of an "old" bucket with capacity pre-dating BigInt change?

williamlardier replied:
-1n, but yes it works with the previous values as well, as we also accept numbers.

williamlardier commented:
ping

ghost commented:
ping

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-188 contains:


* `1.14.17`

* `1.15.8`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.14.18`

* `1.15.8`


Please check the `Fix Version/s` of S3UTILS-188, or the target
branch of this pull request.




*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/1.15/bugfix/S3UTILS-188-support-metrics-above-max-safe` with contents from `bugfix/S3UTILS-188-support-metrics-above-max-safe`
and `development/1.15`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/1.15/bugfix/S3UTILS-188-support-metrics-above-max-safe origin/development/1.15
 $ git merge origin/bugfix/S3UTILS-188-support-metrics-above-max-safe
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/1.15/bugfix/S3UTILS-188-support-metrics-above-max-safe
```





*The following options are set:* **approve**

bert-e commented:
# Build failed



The [build](https://github.com/scality/s3utils/actions?query=branch%3Aw/1.15/bugfix/S3UTILS-188-support-metrics-above-max-safe) for [commit](https://github.com/scality/s3utils/commit/2086ecf4efec9af52bdb0f6b894fd03ec51b6664) did not succeed in branch [w/1.15/bugfix/S3UTILS-188-support-metrics-above-max-safe](https://github.com/scality/s3utils/compare/development/1.15...w/1.15/bugfix/S3UTILS-188-support-metrics-above-max-safe)



*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.14`

* :heavy_check_mark: `development/1.15`



The following branches have **NOT** changed:

* `development/1.13`
* `development/1.4`



Please check the status of the associated issue S3UTILS-188.

Goodbye benzekrimaha.




*The following options are set:* **approve**
